File: 230.kth-smallest-element-in-a-bst.scala
import javax.swing.tree.TreeNode
/*
 * @lc app=leetcode id=230 lang=scala
 *
 * [230] Kth Smallest Element in a BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def kthSmallest(root: TreeNode, k: Int): Int = {
        def reverseOrder(root: TreeNode, ls: List[Int]): List[Int] =
            if (root == null) ls
            else reverseOrder(root.left, root.value :: reverseOrder(root.right, ls))
        reverseOrder(root, Nil).drop(k - 1).head
    }
}
// @lc code=end

File: 230.kth-smallest-element-in-a-bst.py
#
# @lc app=leetcode id=230 lang=python3
#
# [230] Kth Smallest Element in a BST
#

# @lc code=start
# Definition for a binary tree node.
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def loop(root: Optional[TreeNode], cnt: int) -> Optional[int]:
            if root == None:
                return root
            loop(root.left, cnt)
            cnt -= 1
            if cnt == 0:
                return root.val
            loop(root.right, cnt)
        return loop(root, k)


# @lc code=end
